src/control/lib/control/pool.go

```diff
@@ -420,24 +421,71 @@ type (
 	}
 )
 
-func (pi *PoolInfo) MarshalJSON() ([]byte, error) {
-	if pi == nil {
+func (pqr *PoolQueryResp) MarshalJSON() ([]byte, error) {
+	if pqr == nil {
 		return []byte("null"), nil
 	}
 
-	rs, err := system.CreateRankSet(pi.RankSet)
-	if err != nil {
-		return nil, err
+	type Alias PoolQueryResp
+	aux := &struct {
+		EnabledRanks  *[]system.Rank `json:"enabled_ranks"`
+		DisabledRanks *[]system.Rank `json:"disabled_ranks"`
+		*Alias
+	}{
+		Alias: (*Alias)(pqr),
+	}
+
+	if pqr.EnabledRanks != nil {
+		ranks := pqr.EnabledRanks.Ranks()
+		aux.EnabledRanks = &ranks
+	}
+
+	if pqr.DisabledRanks != nil {
+		ranks := pqr.DisabledRanks.Ranks()
+		aux.DisabledRanks = &ranks
 	}
 
-	type Alias PoolInfo
-	return json.Marshal(&struct {
-		RankSet []system.Rank `json:"rank_set"`
+	return json.Marshal(&aux)
+}
+
+func unmarshallRankSet(ranks string) (*system.RankSet, error) {
+	switch ranks {
+	case "":
+		return nil, nil
+	case "[]":
+		return &system.RankSet{}, nil
+	default:
+		return system.CreateRankSet(ranks)
+	}
+}
+
+func (pqr *PoolQueryResp) UnmarshalJSON(data []byte) error {
+	type Alias PoolQueryResp
+	aux := &struct {
+		EnabledRanks  string `json:"enabled_ranks"`
+		DisabledRanks string `json:"disabled_ranks"`
 		*Alias
 	}{
-		RankSet: rs.Ranks(),
-		Alias:   (*Alias)(pi),
-	})
+		Alias: (*Alias)(pqr),
+	}
+
+	if err := json.Unmarshal(data, &aux); err != nil {
+		return err
+	}
+
+	if rankSet, err := unmarshallRankSet(aux.EnabledRanks); err != nil {
+		return err
+	} else {
+		pqr.EnabledRanks = rankSet
+	}
+
+	if rankSet, err := unmarshallRankSet(aux.DisabledRanks); err != nil {
+		return err
+	} else {
+		pqr.DisabledRanks = rankSet
+	}
```

Comment:
Hmm. The problem with this organization is that it will be incompatible with any future work to allow a response that contains both enabled and disabled ranks. I understand that we do not currently support this, but it doesn't seem correct to bake that limitation into the "plumbing", so to speak. Instead, I think that we should keep this middle code agnostic about such details and instead handle them on the server side and on the display side.